File: programs/soap-dispenser/src/instructions/mint.rs
use anchor_lang::prelude::*;
use mpl_bubblegum::{
    program::Bubblegum,
    state::{
        metaplex_anchor::{MplTokenMetadata, TokenMetadata}, TreeConfig,
        COLLECTION_CPI_PREFIX, metaplex_adapter::{MetadataArgs, TokenStandard, Collection, TokenProgramVersion, Creator},
    }};
use spl_account_compression::{Noop, program::SplAccountCompression};
use crate::{states::{Dispenser}, constants::{DISPENSER_PREFIX}, error::SoapCode};

#[derive(Accounts)]
pub struct Mint<'info> {
    #[account(
        seeds=[
            DISPENSER_PREFIX,
            authority.key().as_ref(),
            collection_mint.key().as_ref(),
        ],
        bump,
    )]
    pub dispenser: Box<Account<'info, Dispenser>>,
    #[account(mut)]
    pub payer: SystemAccount<'info>,
    #[account(
        mut,
        seeds = [merkle_tree.key().as_ref()],
        bump,
        seeds::program = bubblegum_program.key()
    )]
    pub tree_authority: Account<'info, TreeConfig>,
    /// CHECK: This account is neither written to nor read from.
    pub leaf_owner: AccountInfo<'info>,
    /// CHECK: This account is neither written to nor read from.
    pub leaf_delegate: AccountInfo<'info>,
    #[account(mut)]
    /// CHECK: unsafe
    pub merkle_tree: UncheckedAccount<'info>,
    #[account(mut)]
    pub authority: SystemAccount<'info>,
    /// CHECK: This account is neither written to nor read from.
    pub tree_delegate: UncheckedAccount<'info>,
    // /// CHECK: This account is neither written to nor read from.
    // pub collection_authority: UncheckedAccount<'info>,
    /// CHECK: Optional collection authority record PDA.
    /// If there is no collecton authority record PDA then
    /// this must be the Bubblegum program address.
    pub collection_authority_record_pda: UncheckedAccount<'info>,
    /// CHECK: This account is checked in the instruction
    pub collection_mint: UncheckedAccount<'info>,
    #[account(mut)]
    pub collection_metadata: Box<Account<'info, TokenMetadata>>,
    /// CHECK: This account is checked in the instruction
    pub edition_account: UncheckedAccount<'info>,
    /// CHECK: This is just used as a signing PDA.
    #[account(
        seeds = [COLLECTION_CPI_PREFIX.as_ref()],
        bump,
        seeds::program = bubblegum_program.key()
    )]
    pub bubblegum_signer: UncheckedAccount<'info>,
    pub bubblegum_program: Program<'info, Bubblegum>,
    pub log_wrapper: Program<'info, Noop>,
    pub compression_program: Program<'info, SplAccountCompression>,
    pub token_metadata_program: Program<'info, MplTokenMetadata>,
    pub system_program: Program<'info, System>,
}

pub fn mint_handler(ctx: Context<Mint>) -> Result<()> {
    let rent: u64 = 10000;
    let pot_balance = **ctx.accounts.payer.to_account_info().lamports.borrow_mut();

    // check if pot has enough funds to transfer back the rent
    if pot_balance.le(&(rent)) {
        return Err(SoapCode::InsufficientFunds)?;
    }

    let clock = Clock::get()?;

    let current_unix_timestamp = clock.unix_timestamp;

    // check if mint is invoked before start date

    match ctx.accounts.dispenser.start_date {
        Some(s_d) => if current_unix_timestamp.le(&s_d) {
            return Err(SoapCode::MintNotStarted)?;
        }
        None => {}
    }

    // check if mint is invoked after end date
    match  ctx.accounts.dispenser.end_date {
        Some(e_d) => if current_unix_timestamp.ge(&e_d) {
            return Err(SoapCode::MintEnded)?;
        }
        None => {}
    }

    let cpi_accounts = mpl_bubblegum::cpi::accounts::MintToCollectionV1 {
        tree_authority: ctx.accounts.tree_authority.to_account_info(),
        leaf_owner: ctx.accounts.leaf_owner.to_account_info(),
        leaf_delegate: ctx.accounts.leaf_delegate.to_account_info(),
        merkle_tree:  ctx.accounts.merkle_tree.to_account_info(),
        payer: ctx.accounts.payer.to_account_info(),
        tree_delegate: ctx.accounts.dispenser.to_account_info(),
        collection_authority: ctx.accounts.dispenser.to_account_info(),
        collection_authority_record_pda: ctx.accounts.collection_authority_record_pda.to_account_info(),
        collection_mint: ctx.accounts.collection_mint.to_account_info(),
        collection_metadata: ctx.accounts.collection_metadata.to_account_info(),
        edition_account: ctx.accounts.edition_account.to_account_info(),
        bubblegum_signer: ctx.accounts.bubblegum_signer.to_account_info(),
        log_wrapper: ctx.accounts.log_wrapper.to_account_info(),
        compression_program: ctx.accounts.compression_program.to_account_info(),
        token_metadata_program: ctx.accounts.token_metadata_program.to_account_info(),
        system_program: ctx.accounts.system_program.to_account_info(),
    };

    let dispenser_seeds = &[
        &DISPENSER_PREFIX[..],
        ctx.accounts.authority.key.as_ref(),
        ctx.accounts.collection_mint.key.as_ref(),
        &[*ctx.bumps.get("dispenser").unwrap()],
    ];
    
    mpl_bubblegum::cpi::mint_to_collection_v1(CpiContext::new_with_signer(
        ctx.accounts.bubblegum_program.to_account_info(),
        cpi_accounts,
        &[&dispenser_seeds[..]]
    ), MetadataArgs { 
        name: ctx.accounts.dispenser.soap_details.name.clone(),
        symbol: ctx.accounts.dispenser.soap_details.symbol.clone(),
        uri: ctx.accounts.dispenser.soap_details.uri.clone(),
        seller_fee_basis_points: ctx.accounts.dispenser.soap_details.seller_fee_basis_points,
        primary_sale_happened: false,
        is_mutable: true,
        edition_nonce: None,
        token_standard: Some(TokenStandard::NonFungible),
        collection: Some(Collection { 
            verified: false, 
            key: ctx.accounts.dispenser.collection_mint.key() 
        }),
        uses: None,
        token_program_version: TokenProgramVersion::Original,
        creators: vec![Creator {
            address: ctx.accounts.dispenser.creator.key(),
            share: 100,
            verified: false
        }],
    })?;
    Ok(())
}

File: scripts/mint.ts
import { Metaplex, keypairIdentity } from "@metaplex-foundation/js";
import {
  PROGRAM_ID as BUBBLEGUM_PROGRAM_ID,
  MetadataArgs,
  TokenProgramVersion,
  TokenStandard,
} from "@metaplex-foundation/mpl-bubblegum";
import { PROGRAM_ID as TOKEN_METADATA_PROGRAM_ADDRESS } from "@metaplex-foundation/mpl-token-metadata";
import {
  SPL_ACCOUNT_COMPRESSION_PROGRAM_ID,
  SPL_NOOP_PROGRAM_ID,
} from "@solana/spl-account-compression";
import {
  Keypair,
  PublicKey,
  SystemProgram,
  sendAndConfirmTransaction,
} from "@solana/web3.js";
import {
  Workspace,
  connection,
  findDispenserPda,
  findPotPda,
  findTreeAuthorityPda,
  getAssetId,
  getTokenProgramVersion,
  getTokenStandard,
  network,
  sleep,
} from ".";
import { initializeKeypair } from "./initializeKeypair";
import { Network, getUrls } from "./networks";
import { WrapperConnection } from "./WrapperConnection";

export const mint = async ({
  authority,
  metaplex,
  collectionMint,
  buyer,
  payer,
}: {
  authority: Keypair;
  metaplex: Metaplex;
  collectionMint: PublicKey;
  buyer: PublicKey;
  payer: Keypair;
}) => {
  try {
    const { program, programId, provider, connection } = new Workspace(
      authority
    );

    const dispenser = findDispenserPda(authority.publicKey, collectionMint);
    const pot = findPotPda(authority.publicKey, dispenser, collectionMint);

    const dispenserAccount = await program.account.dispenser.fetch(dispenser);
    console.log(dispenserAccount);
    const merkleTree = dispenserAccount.tree.merkleTree;
    const treeAuthority = dispenserAccount.tree.authority;
    // const treeAuthority = findTreeAuthorityPda(merkleTree);
    const metadata = metaplex.nfts().pdas().metadata({ mint: collectionMint });
    const masterEdition = metaplex
      .nfts()
      .pdas()
      .masterEdition({ mint: collectionMint });
    const collectionAuthorityRecord = metaplex
      .nfts()
      .pdas()
      .collectionAuthorityRecord({
        mint: collectionMint,
        collectionAuthority: dispenser,
      });

    const [bubblegumSigner] = PublicKey.findProgramAddressSync(
      [Buffer.from("collection_cpi", "utf8")],
      BUBBLEGUM_PROGRAM_ID
    );
    const mintTx = await program.methods
      .mint()
      .accounts({
        dispenser,
        payer: payer.publicKey,
        treeAuthority,
        leafOwner: buyer,
        leafDelegate: buyer,
        merkleTree: merkleTree,
        authority: authority.publicKey,
        treeDelegate: dispenser,
        collectionAuthorityRecordPda: collectionAuthorityRecord,
        collectionMint: collectionMint,
        collectionMetadata: metadata,
        editionAccount: masterEdition,
        bubblegumSigner,
        bubblegumProgram: BUBBLEGUM_PROGRAM_ID,
        logWrapper: SPL_NOOP_PROGRAM_ID,
        tokenMetadataProgram: TOKEN_METADATA_PROGRAM_ADDRESS,
        compressionProgram: SPL_ACCOUNT_COMPRESSION_PROGRAM_ID,
        systemProgram: SystemProgram.programId,
      })
      .signers([payer])
      .transaction();

    const sig = await sendAndConfirmTransaction(connection, mintTx, [payer], {
      commitment: "confirmed",
    });

    console.log("Mint:", getUrls(Network[network], sig, "tx").explorer);
  } catch (error) {
    throw error;
  }
};

const main = async () => {
  try {
    const authority = await initializeKeypair(connection, "soap_creator_1");
    const buyer = await initializeKeypair(connection, "buyer_1");
    const payer = await initializeKeypair(connection, "fund_wallet");
    const metaplex = Metaplex.make(connection).use(keypairIdentity(authority));
    const metadataArgs: MetadataArgs = {
      name: "First cNFT",
      symbol: "cNFT",
      uri: "https://arweave.net/mo4NBHmhuCt9ZjJ6jykMgKw-te0uTdgDBkBjVAJ-v20",
      sellerFeeBasisPoints: 500,
      primarySaleHappened: false,
      isMutable: true,
      creators: [
        {
          address: authority.publicKey,
          share: 100,
          verified: false,
        },
      ],
      collection: {
        key: new PublicKey(process.argv[2]),
        verified: false,
      },
      editionNonce: null,
      tokenProgramVersion: TokenProgramVersion.Original,
      tokenStandard: TokenStandard.NonFungible,
      uses: null,
    };
    await mint({
      authority,
      metaplex,
      buyer: buyer.publicKey,
      collectionMint: new PublicKey(process.argv[2]),
      payer,
    });
  } catch (error) {
    console.log(error);
  }
};

main();
